### PR TITLE
New semantics for equality on IP ranges

### PR DESCRIPTION
**Pull request by cdisselkoen**
## Description of changes

Changes the semantics of equality on IP ranges.  For example, `ip("192.168.0.1/24") == ip("192.168.0.3/24")` was previously `true` and is now `false`.

Does not affect equality of single IP addresses, and does not change the behavior of `.isInRange()` at all.

Includes a lot more tests for edge-cases of IP semantics, including many tests for equality with the same or different subnets, equality between a single address and a range, and `.isInRange()` on a wider variety of edge cases.

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)

I may not make these updates myself, but I'm in contact with Cedar maintainers who are preparing to make these changes.

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cedar-policy-core/src/extensions/ipaddr.rs [hunk 1]
```diff
@@ -46,9 +45,11 @@ const ADVICE_MSG: &str = "Maybe you forgot to apply the `ip` constructor?";
 
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
 struct IPAddr {
-    /// the actual address, represented internally as an `IpNet`.
-    /// A single address is represented as an `IpNet` containing only one address.
-    addr: ipnet::IpNet,
+    /// the actual address, without subnet
```
Comment (shaobo-he-aws) [inline cedar-policy-core/src/extensions/ipaddr.rs:48]: Should we rename this struct to `IpNet`?

File: cedar-policy-core/src/extensions/ipaddr.rs [hunk 2]
```diff
@@ -61,34 +62,25 @@ impl IPAddr {
...
-            Ok(singleaddr) => ipnet::IpNet::from(singleaddr),
-            Err(e1) => match parse_ipnet(str.as_ref()) {
-                Ok(net) => net.trunc(),
-                Err(e2) => return Err(format!(
-                    "error parsing IP address from string. Failed to parse as single address: `{}`. Failed to parse as subnet: `{}`", e1, e2)),
-            }
-        };
-        Ok(Self { addr })
+        // Delegate to `FromStr` implementation
+        str.as_ref().parse()
```
Comment (shaobo-he-aws) [inline cedar-policy-core/src/extensions/ipaddr.rs:73]: Do we need this function? Why not just use the implementation of `FromStr`?

File: cedar-policy-core/src/extensions/ipaddr.rs [hunk 3]
```diff
@@ -97,21 +89,80 @@ impl IPAddr {
...
-        // If an IpNet's addresses are multicast addresses, calling `is_in_range()` over it and its associated net above should evaluate to true
-        // The implementation uses the property that if `ip1/prefix1` is in range `ip2/prefix2`, then `ip1` is in `ip2/prefix2` and `prefix1 >= prefix2`
-        self.addr.addr().is_multicast()
-            && self.addr.prefix_len() >= if self.is_ipv4() { 4 } else { 8 }
+        // If an ip range's addresses are multicast addresses, calling
+        // `is_in_range()` over it and its associated net above should evaluate
+        // to true
+        // The implementation uses the property that if `ip1/prefix1` is in range
+        // `ip2/prefix2`, then `ip1` is in `ip2/prefix2` and `prefix1 >= prefix2`
+        self.addr.is_multicast() && self.subnet >= if self.is_ipv4() { 4 } else { 8 }
```
Comment (shaobo-he-aws) [inline cedar-policy-core/src/extensions/ipaddr.rs:103]: Should `>=` here be just `==`?

**Review by emina — APPROVED** [review]

**Review by andrewmwells-amazon — APPROVED** [review]

**Review by andrewmwells-amazon — CHANGES_REQUESTED** [review]
I think we should more seriously consider splitting networks and addresses into different types.

**Comment by emina** [comment]
Splitting networks and addresses would be an invasive breaking change.  We’d break the schema of every user of the IP extension, because they’d have to change their IP attribute types to differentiate between addresses and ranges.

**Review by aaronjeline — APPROVED** [review]

**Review by andrewmwells-amazon — APPROVED** [review]

**Review by shaobo-he-aws — APPROVED** [review]

**Review by andrewmwells-amazon — APPROVED** [review]

**Review by mwhicks1 — APPROVED** [review]

**Review by aaronjeline — APPROVED** [review]
